azurerm/resource_arm_policy_set_definition.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/terraform-providers/terraform-provider-azurerm/azurerm/utils"
+)
+
+func resourceArmPolicySetDefinition() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceArmPolicySetDefinitionCreateUpdate,
+		Update: resourceArmPolicySetDefinitionCreateUpdate,
+		Read:   resourceArmPolicySetDefinitionRead,
+		Delete: resourceArmPolicySetDefinitionDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"policy_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					string(policy.TypeBuiltIn),
+					string(policy.TypeCustom),
+					string(policy.TypeNotSpecified),
+				}, true)},
+
+			"display_name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"metadata": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ValidateFunc:     validation.ValidateJsonString,
+				DiffSuppressFunc: structure.SuppressJsonDiff,
+			},
+
+			"parameters": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ValidateFunc:     validation.ValidateJsonString,
+				DiffSuppressFunc: structure.SuppressJsonDiff,
+			},
+
+			"policy_definitions": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ValidateFunc:     validation.ValidateJsonString,
+				DiffSuppressFunc: policyDefinitionsDiffSuppressFunc,
+			},
+		},
+	}
+}
+
+func policyDefinitionsDiffSuppressFunc(k, old, new string, d *schema.ResourceData) bool {
+	var oldPolicyDefinitions []policy.DefinitionReference
+	errOld := json.Unmarshal([]byte(old), &oldPolicyDefinitions)
+	if errOld != nil {
+		return false
+	}
+
+	var newPolicyDefinitions []policy.DefinitionReference
+	errNew := json.Unmarshal([]byte(old), &newPolicyDefinitions)
+	if errNew != nil {
+		return false
+	}
+
+	return reflect.DeepEqual(oldPolicyDefinitions, newPolicyDefinitions)
+}
+
+func resourceArmPolicySetDefinitionCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).policySetDefinitionsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	policyType := d.Get("policy_type").(string)
+	displayName := d.Get("display_name").(string)
+	description := d.Get("description").(string)
+
+	properties := policy.SetDefinitionProperties{
+		PolicyType:  policy.Type(policyType),
+		DisplayName: utils.String(displayName),
+		Description: utils.String(description),
+	}
+
+	if metaDataString := d.Get("metadata").(string); metaDataString != "" {
+		metaData, err := structure.ExpandJsonFromString(metaDataString)
+		if err != nil {
+			return fmt.Errorf("unable to parse metadata: %s", err)
```

Comment:
Would this be a better error message?
```suggestion
			return fmt.Errorf("unable to expand metadata json: %s", err)
```